numpyro/distributions/continuous.py

```diff
@@ -1,441 +1,373 @@
-
-    def _entropy(self):
-        return 0.5 * np.log(np.pi / 2.0) + 0.5
-
-
-class lognorm_gen(jax_continuous):
-    arg_constraints = {'s': constraints.positive}
-    _support_mask = constraints.positive
-
-    def _rvs(self, s):
-        # TODO: use upstream implementation when available
-        return np.exp(s * random.normal(self._random_state, shape=self._size))
-
-    def _pdf(self, x, s):
-        # lognorm.pdf(x, s) = 1 / (s*x*sqrt(2*pi)) * exp(-1/2*(log(x)/s)**2)
-        return np.exp(self._logpdf(x, s))
-
-    def _logpdf(self, x, s):
-        return np.where(x != 0,
-                        -np.log(x) ** 2 / (2 * s ** 2) - np.log(s * x * np.sqrt(2 * np.pi)),
-                        -np.inf)
-
-    def _cdf(self, x, s):
-        return norm._cdf(np.log(x) / s)
-
-    def _logcdf(self, x, s):
-        return norm._logcdf(np.log(x) / s)
-
-    def _ppf(self, q, s):
-        return np.exp(s * norm._ppf(q))
-
-    def _sf(self, x, s):
-        return norm._sf(np.log(x) / s)
-
-    def _logsf(self, x, s):
-        return norm._logsf(-np.log(x) / s)
-
-    def _stats(self, s):
-        p = np.exp(s * s)
-        mu = np.sqrt(p)
-        mu2 = p * (p - 1)
-        g1 = np.sqrt(p - 1) * (2 + p)
-        g2 = np.polyval([1, 2, 3, 0, -6.0], p)
-        return mu, mu2, g1, g2
-
-    def _entropy(self, s):
-        return 0.5 * (1 + np.log(2 * np.pi) + 2 * np.log(s))
-
-
-class norm_gen(jax_continuous):
-    _support_mask = constraints.real
-
-    def _rvs(self):
-        return random.normal(self._random_state, shape=self._size)
-
-    def _pdf(self, x):
-        # norm.pdf(x) = exp(-x**2/2)/sqrt(2*pi)
-        return np.exp(-x**2 / 2.0) / np.sqrt(2 * np.pi)
-
-    def _logpdf(self, x):
-        return -(x ** 2 + np.log(2 * np.pi)) / 2.0
-
-    def _cdf(self, x):
-        return ndtr(x)
-
-    def _logcdf(self, x):
-        return log_ndtr(x)
-
-    def _sf(self, x):
-        return ndtr(-x)
-
-    def _logsf(self, x):
-        return log_ndtr(-x)
-
-    def _ppf(self, q):
-        return ndtri(q)
-
-    def _isf(self, q):
-        return -ndtri(q)
-
-    def _stats(self):
-        return 0.0, 1.0, 0.0, 0.0
-
-    def _entropy(self):
-        return 0.5 * (np.log(2 * np.pi) + 1)
-
+from numpyro.distributions.constraints import AbsTransform, AffineTransform, ExpTransform
+from numpyro.distributions.distribution import Distribution, TransformedDistribution
+from numpyro.distributions.util import get_dtypes, promote_shapes, standard_gamma
+
+
+class Beta(Distribution):
+    arg_constraints = {'concentration1': constraints.positive, 'concentration0': constraints.positive}
+    support = constraints.unit_interval
+
+    def __init__(self, concentration1, concentration0, validate_args=None):
+        batch_shape = lax.broadcast_shapes(np.shape(concentration1), np.shape(concentration0))
+        self.concentration1 = np.broadcast_to(concentration1, batch_shape)
+        self.concentration0 = np.broadcast_to(concentration0, batch_shape)
```

Comment:
`promote_shapes` isn't sufficient here since we have to stack `conc1`, `conc0` which don't  broadcast internally, so a full broadcasting is unfortunately needed. I think we can have it return the full batch shape too. `lax.broadcast_shapes` should be really cheap though, so I'm not sure if its worth the effort.

Reply:
Ah, I thought that `promote_shapes` returns broadcasted tensors. So many comments from me are invalid. >"< Sorry about that!

Reply:
I think the larger issue is that in numpy/jax, we don't have something like `torch.expand` which returns a broadcasted array but doesn't modify the underlying memory (just returns a view). Therefore, we shouldn't naively do full broadcasting unless really needed, like in this case. `promote_shapes` just reshapes the arrays and does not need to copy the memory.

Reply:
It is good to know! Thanks @neerajprad !